service/dap/server.go

```diff
@@ -171,6 +171,7 @@ func (s *Server) signalDisconnect() {
 	}
 	if s.binaryToRemove != "" {
 		gobuild.Remove(s.binaryToRemove)
```

Comment:
The test is flaky. 

What I observed is the debugger.Debugger.Detach is called (most likely as a result of  handling `onDisconnectRequest`), but the process already exited so [the actual `Debugger.detach` isn't called](https://github.com/go-delve/delve/blob/c068861f95ec570fb7ea93638c87338f25c96af3/service/debugger/debugger.go#L418) .  But the test still fails.

I added 10ms Sleep before calling gobuild.Remove, and that made the test pass with -count=100. :-(

I think the race is somewhere deep down.

Reply:
Debugger.Detach is called twice actually (by disconnect and by server stop). This is by design because you could end up stopping as a result of an explicit disconnect or because of an error or an interrupt (ctrl-c), skipping the disconnect request.

In both cases, here the Detach method does not call the underlying target's detach because the target has terminated by the time it is called. This is expected for a test where the program runs to completion. We should be able to test such cases and not always disconnect tests forcefully in the middle. When a process exits, a terminate event is sent. And that triggers a disconnect request on the editor side. This is per DAP spec. So the sequence looks right to me here.

I would expect the debugger backend to release the binary when the program terminates. Does it do that? Does it do that too slowly somehow, so we end up trying to remove while the file is still in use? And if that is the case, this is not a flaky test problem, but a general problem because users could see the "could not remove" messages in the wild when the server cleans up when the program runs to completion.

Reply:
@polinasok sorry if it's not clear - I didn't mean the sequence was wrong, but meant to reply to @aarzilli's previous comment about making sure Detach was called. And the test doesn't always fail and that's why I said 'flaky'. 

I agree that it looks like the problem is the debugger isn't releasing the binary as fast as we thought.

Reply:
> I would expect the debugger backend to release the binary when the program terminates. Does it do that?

In theory it should be done by postExit, it is possible that we missed a spot. 

> I would expect the debugger backend to release the binary when the program terminates. Does it do that?

It is possible that this is the windows thing that we never figured out. It is possible that the thing that's keeping the file occupied is not delve but windows, because the program was executed recently.

Reply:
So `postExit` does get called here. It must be Windows then :(

Here is the stacktrace:
```
github.com/go-delve/delve/pkg/proc/native.(*nativeProcess).postExit(0xc000a20100)
  Z:/buildAgent/work/3aca8f960d346a9c/pkg/proc/native/proc.go:319 +0x8a
github.com/go-delve/delve/pkg/proc/native.(*nativeProcess).trapWait(0xc000a20100, 0xffffffffffffffff, 0xc000965048, 0x1, 0x1)
  Z:/buildAgent/work/3aca8f960d346a9c/pkg/proc/native/proc_windows.go:378 +0x135
github.com/go-delve/delve/pkg/proc/native.(*nativeProcess).ContinueOnce(0xc000a20100, 0xc0002da000, 0x1, 0x1, 0xc000965170, 0x145b78)
  Z:/buildAgent/work/3aca8f960d346a9c/pkg/proc/native/proc.go:246 +0x15b
github.com/go-delve/delve/pkg/proc.(*Target).Continue(0xc00045fb00, 0x0, 0x0)
  Z:/buildAgent/work/3aca8f960d346a9c/pkg/proc/target_exec.go:72 +0x1ff
github.com/go-delve/delve/service/debugger.(*Debugger).Command(0xc0007486e0, 0xc0009658f0, 0x0, 0x0, 0x0)
  Z:/buildAgent/work/3aca8f960d346a9c/service/debugger/debugger.go:927 +0x1194
github.com/go-delve/delve/service/dap.(*Server).doCommand(0xc0005f4080, 0x474aeb, 0x8)
```

Reply:
Isn't it better to retry to Remove a couple of times in case of failure if a bit of delay could make it succeed? It's annoying to see delve leaving binaries.

Reply:
https://github.com/go-delve/delve/pull/2351